yaml.go

```diff
@@ -112,6 +127,15 @@ func (dec *Decoder) KnownFields(enable bool) {
 	dec.knownFields = enable
 }
 
+// DecoderHook is called for every leaf node and receives the node itself and
+// the path leading up to the node.
+type DecoderHook func(path []string, node *Node)
+
+// WithHook adds a DecoderHook that gets called for every leaf node.
+func (dec *Decoder) WithHook(h DecoderHook) {
```

Comment:
IIUC, this hook will be called as nodes are being parsed (i.e. we won't wait for the whole yaml to be parsed to execute the hooks)? In other words, it's possible that parsing fails, but the hook may still have been executed?